### ISSUE TITLE
Tailored web views for Cartography

### ISSUE THREAD

**Issue by achantavy** [lyft/cartography#413]
*Description*:
> Describe your idea.  Please be detailed.  If a feature request, please
describe the desired behavior, what scenario it enables, and how it
would be used.

Cartography enables powerful correlations between technical assets by representing them in a graph database, but there is a steep curve in learning the Cypher query language to extract these insights. 

To make this data useful for more people, we would like to expose this graph data in a tailored web view.

In a nutshell we want to do this:
![image](https://user-images.githubusercontent.com/46503781/94953854-14c89b80-049d-11eb-89a3-cd680fbb6a87.png)

In the spirit of [Hacktoberfest](https://hacktoberfest.digitalocean.com/), I am trying to avoid being too prescriptive in this feature idea to hopefully encourage creative solutions :-).

**Comment by other019** [comment]
So if I understand this correctly its about creating small standalone project for runing predefined queries?

**Comment by achantavy** [comment]
Correct. It could live in a folder called web-view/frontend/whatever-you-want-to-call-it here: https://github.com/lyft/cartography/tree/master/cartography

**Comment by other019** [comment]
So, I think I try to do this. Any preferences to frontend? React or Vue?

**Comment by achantavy** [comment]
Up to you! :D

**Comment by achantavy** [comment]
Just wanted to add that we're willing to accept more than one frontend as I imagine our users will have different preferences for their own environments.

**Comment by adam-cowley** [comment]
I'm happy to pair/contribute to this and livestream the process.  Have you got an example config file?  What kind of charts are you looking for?

**Comment by achantavy** [comment]
Awesome.

I think a query file format should look like this: https://github.com/marco-lancini/cartography-queries/blob/master/queries/queries.json

Of those, I'm particularly interested in stuff with IAM permissions like this one: https://github.com/marco-lancini/cartography-queries/blob/085014e3f6ed3f9254a8aff0bf7a84109b56ce22/queries/queries.json#L260-L271

Example, say we have an AWS admin role. I want a quick list of which AWS users (or Okta users) who can assume that role without needing to look up the query and also without needing to ask the one dev who knows Cypher at the company to run that query for me haha.

As far as visualizations go, it might be neat to show grouping of assets by account, like for example "which AWS accounts have more EC2 instances compared to the others". This is just one idea though; there's lots of directions to take this but in general grouping stuff up is helpful in visualizations.

Please do let us know when you decide to livestream :).

**Comment by adam-cowley** [comment]
OK cool, these all look like table outputs to me.  Do you envisage UI forms to aggregate on key/turn into a chart?  I've been thinking about how I could do a Mixpanel type UI based on Cypher queries for a while.

@other019 there's [use-neo4j](https://github.com/adam-cowley/use-neo4j) for React and [vue-neo4j](https://github.com/adam-cowley/vue-neo4j) for Vue2 which would help speed up development if you go down the [Graph App](https://neo4j.com/developer/graph-app-development/) route.

**Comment by achantavy** [comment]
> Do you envisage UI forms to aggregate on key/turn into a chart? I've been thinking about how I could do a Mixpanel type UI based on Cypher queries for a while.

This sounds like a great bonus.  My imagination in this area is very limited haha, I'll let you make the decisions here. My requirement is simple: I'm only envisioning a world where people can look at this page without needing to ask me to write queries :-). Or if I can quickly write up a query once, define in a config file or page somewhere, and then have its results quickly presented in some view then that's awesome already.

>  Have you got a demo dataset? I don't think my two dev servers in AWS will do this any justice.

One hacky way to get test data in your local graph is to comment out this line in conftest.py: https://github.com/lyft/cartography/blob/4fdc8518f546d8329669f23d1fe928a2158c52d9/tests/integration/conftest.py#L12
and then run `make test_integration`. This will complain and give you all kinds of error messages but your local Neo4j instance will now be full of test data from our integration tests.

**Comment by other019** [comment]
I'm sorry but I need to give up for some time. To much duties in my  job. I'm sorry

**Comment by krisek** [comment]
Has anybody experimented with [Drivine](https://drivine.org)? It seems to be pretty much what this issue is about.

**Comment by adam-cowley** [comment]
> and then run `make test_integration`. This will complain and give you all kinds of error messages but your local Neo4j instance will now be full of test data from our integration tests.

I've just tried to pick this up this morning.  I'm pretty inexperienced with python but I've installed the test-dependencies through pip and even cartographer.  From a fresh install I get errors in 7 files telling me:

```
E   ModuleNotFoundError: No module named 'okta.framework'
```

I've tried running `pip install okta` without success.  Is there a particular version of okta or any additional dependencies that I should be installing?

**Comment by achantavy** [comment]
Hi @adam-cowley, I saw you posted a question earlier asking for help, were you able to figure it out? I think you need to pull the latest revision from the Cartography master branch because of #424 and #425. If you're having trouble I'll put together an out-of-band release so you can pull the latest fix from Pip.

**Comment by adam-cowley** [comment]
I got it working in the end so deleted the message but thanks for the reply.  

I've been working on something recently that was partly inspired by this thread that may be useful - a [Graph App](https://neo4j.com/developer/graph-apps/) for [Neo4j Desktop](https://neo4j.com/developer/neo4j-desktop/) that allows users with domain knowledge but without cypher experience to build up queries and then attach them to reports ([screenshot here](https://twitter.com/adamcowley/status/1325751570588381184)).

I'm currently working though the [different chart types](https://nivo.rocks/) that might be useful but for now you can display data in tables, line charts, bar charts and radar charts.  I'm thinking maybe a bubble chart would be useful for servers/groups?

You can install it by downloading [this .tgz](https://drive.google.com/file/d/1k9gTnPFuDwZSjPRWwvMEJK0ovBZECM1H/view) and dragging or pasting the file location into the Install form at the bottom of the graph app in Neo4j Desktop.  It'll give you a graph app called GraphPanel (working title) with the React logo.  Once you open it up you have links in the top right to Query Builder and Dashboards. 

The app saves two json files to Neo4j Desktop which can be shared with colleagues so colleagues can view the same reports/data.

Have a click around and let me know what you think.

**Comment by achantavy** [comment]
Hi @adam-cowley, thanks for sharing, I'll have to play with your tarball soon now that things have quieted down a bit around the office for the holidays :). Do you have this code in a Github repo by any chance?

**Comment by adam-cowley** [comment]
Sure, [the repo is here](https://github.com/neo4j-labs/charts) or you can run/install it from install.graphapp.io.  Let me know what you think!

**Comment by adam-cowley** [comment]
Oh... also [this blog post on the Neo4j Developer blog](https://medium.com/neo4j/creating-charts-from-your-graphs-2f5b4e86fd6c) to walk you through how to use it.
